crates/rethnet_rpc_server/src/lib.rs

```diff
@@ -67,67 +74,200 @@ where
     }
 }
 
+/// returns the state root in effect BEFORE setting the block context, so that the caller can
+/// restore the context to that state root.
+async fn set_block_context<T>(
+    state: &StateType,
+    block_spec: BlockSpec,
+) -> Result<B256, ResponseData<T>> {
+    let previous_state_root =
+        state.rethnet.read().await.state_root().map_err(|_| {
+            ResponseData::new_error(0, "Failed to retrieve previous state root", None)
+        })?;
+    if block_spec == BlockSpec::Tag(String::from("latest")) {
+        // do nothing
+        Ok(previous_state_root)
+    } else {
+        state
+            .rethnet
+            .write()
+            .await
+            .set_block_context(
+                &KECCAK_EMPTY,
+                Some(match block_spec {
+                    BlockSpec::Number(n) => Ok(n),
+                    BlockSpec::Tag(tag) => {
+                        if let Some(fork_client) = state.fork_client.clone() {
+                            if let Some(block_number) =
+                                (*fork_client)
```

Comment:
If you query this, you'll potentially get a block number that's older than the latest block number supported in the `ForkState`.

If I'm not mistaken, both Rethnet & Hardhat's forked states make the block number immutable?

I had a look at the spec of `eth_getBlockByNumber` and our own implementation in Hardhat and only 5 different tags are supported. You can model that using an enum and make it easier on yourself:

https://github.com/NomicFoundation/hardhat/blob/main/packages/hardhat-core/src/internal/core/jsonrpc/types/input/blockTag.ts#L15

Reply:
> If you query this, you'll potentially get a block number that's older than the latest block number supported in the `ForkState`.

After some discussion, we realized that this should have said NEWER, not older.

Reply:
Addressed in https://github.com/NomicFoundation/hardhat/pull/3998/commits/1b2980892569f3bd219d2546635a3939c4393849